rclpy/rclpy/node.py

```diff
@@ -248,14 +329,29 @@ def set_parameters(self, parameter_list: List[Parameter]) -> List[SetParametersR
         For each successfully set parameter, a :class:`ParameterEvent` message is
         published.
 
+        If setting a parameter fails due to not being declared, then the
+        parameters which have already been set will stay set, and no attempt will
+        be made to set the parameters which come after.
+
         :param parameter_list: The list of parameters to set.
         :return: A list of SetParametersResult messages.
         """
+        return self._set_parameters(parameter_list, self.set_parameters_atomically)
+
+    def _set_parameters(self, parameter_list: List[Parameter], setter_func=None, raise_on_failure=False) -> List[SetParametersResult]:
+        """
+        Set parameters for the node, without checking whether they were declared beforehand or not.
+        Executes the callback for each parameter in the list.
+        """
+        if setter_func is None:
+            setter_func = self._set_parameters_atomically
+
         results = []
         for param in parameter_list:
-            if not isinstance(param, Parameter):
-                raise TypeError("parameter must be instance of type '{}'".format(repr(Parameter)))
-            results.append(self.set_parameters_atomically([param]))
+            result = setter_func([param])
+            if raise_on_failure and not result.successful:
```

Comment:
What is the imagined use case for `raise_on_failure`? I don't see it being used anywhere.

Reply:
Good catch; this is used in `declare_parameters`. The callback may reject the parameter values, and in this case it should raise an exception. Added in https://github.com/ros2/rclpy/pull/325/commits/4440e852ddf7bdfc84bcc09ba9ace73c94769fa8.